cmake/modules/SetRPATH.cmake

```diff
@@ -0,0 +1,15 @@
+set(CMAKE_SKIP_BUILD_RPATH FALSE)
+set(CMAKE_BUILD_WITH_INSTALL_RPATH FALSE)
+
+# Append non-standard external dependency directories
+set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
```

Comment:
libcmark has no external dependencies, so this is unnecessary.

Reply:
Of course `libcmark` has external dependencies.

<pre>
$ <a href="https://man7.org/linux/man-pages/man1/ldd.1.html">ldd</a> <b>/path/to/installed/files</b>/bin/cmark
        linux-gate.so.1 (0xb7ef2000)
        libcmark.so.0.31.0 => /path/to/installed/files/lib/libcmark.so.0.31.0 (0xb7e95000)
        libc.so.6 => <b>/usr/lib/libc.so.6</b> (0xb7cbc000)
        /lib/ld-linux.so.2 => <b>/usr/lib/ld-linux.so.2</b> (0xb7ef3000)
</pre>

For example:
<pre>
cd "$repo"
d=/tmp/custom
p='
    {print}
    /^include\(GNUInstallDirs\)$/ {
      print("link_directories(\"'"$d"'\")")
    }
'
f=CMakeLists.txt; t=$f.tmp; <"$f" awk >"$t" "$p" && mv -f "$t" "$f"
git diff | cat
echo '================='
i='/path/to/installed/files'
make INSTALL_PREFIX="$i" install | grep 'Set runtime path'
</pre>

Here are the results of running those commands:

<pre>
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 054f8d4..22d9506 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -24,6 +24,7 @@ include(CMakePackageConfigHelpers)
 include(CTest)
 include(GenerateExportHeader)
 include(GNUInstallDirs)
<b>+link_directories("/tmp/custom")</b>
 
 if(NOT MSVC OR CMAKE_HOST_SYSTEM_NAME STREQUAL Windows)
   set(CMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_NO_WARNINGS ON)
=================
-- Set runtime path of "/path/to/installed/files/bin/cmark" to "/path/to/installed/files/lib:<b>/tmp/custom</b>"
-- Set runtime path of "/path/to/installed/files/lib/libcmark.so.0.31.0" to "/path/to/installed/files/lib:<b>/tmp/custom</b>"
</pre>

Then:

<ul>
<li>
<b>Default to System Library</b>
<pre>
$ ldd "$i"/bin/cmark
        linux-gate.so.1 (0xb7ef2000)
        libcmark.so.0.31.0 => /path/to/installed/files/lib/libcmark.so.0.31.0 (0xb7e95000)
        libc.so.6 => <b>/usr/lib</b>/libc.so.6 (0xb7cbc000)
        /lib/ld-linux.so.2 => /usr/lib/ld-linux.so.2 (0xb7ef3000)
</pre>
</li>
<li>
<b>Custom Library</b>
<pre>
$ mkdir -p "$d"
$ cp -a -t "$d" /usr/lib/libc.so.6 /usr/lib/libc-2.29.so
$ ldd "$i"/bin/cmark
        linux-gate.so.1 (0xb7ef2000)
        libcmark.so.0.31.0 => /path/to/installed/files/lib/libcmark.so.0.31.0 (0xb7e95000)
        libc.so.6 => <b>/tmp/custom</b>/libc.so.6 (0xb7cbc000)
        /lib/ld-linux.so.2 => /usr/lib/ld-linux.so.2 (0xb7ef3000)
</pre>
</li>
</ul>

In other words:

* By simply adding `link_directories("/tmp/custom")` to the `CMakeLists.txt`, the directory `/tmp/custom` got added to the `rpath`.

* The only reason this works is because `set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)`.

It's not unreasonable to think that at some point, in the future, an external dependency might be possible; maybe Nvidia will release a driver for offloading markdown parsing to the GPU. :-P

Reply:
Well, obviously the C library is a dependency but it doesn't seem useful to support this use case. Maybe we should also consider to only set the INSTALL_RPATH target property for the cmark executable.

Reply:
* Shared libraries have dependencies, too.

        $ ldd "$i"/lib/libcmark.so
                linux-gate.so.1 (0xb7fa3000)
                libc.so.6 => /tmp/custom/libc.so.6 (0xb7da3000)
                /usr/lib/ld-linux.so.2 (0xb7fa4000)

* **However**, it's straightforward enough to set only `cmark`&rsquo;s `rpath` to an unorthodox installation path, while still allowing `cmake` to set `libcmark`&rsquo;s `rpath` if it deems necessary (as described above); then, in most cases (pretty much *always*), `libcmark`&rsquo;s `rpath` will be the empty string, even when installed in an unorthodox location.

* The goal isn&rsquo;t to support *that* use case.

    (However, it would indeed *now* be supported; I mean, maybe someone wants to build against [*musl*](https://www.musl-libc.org/) or something, and has put everything in a separate, non-standard hierarchy.)

    Rather, the goal is to support a *logical policy* in general, namely:

    > If `cmake` is *ever* aware of an external dependency (or search directory) that is not covered by the standard locations, then `cmake` should add the relevant path to the `rpath`.

    That sounds like a sensible policy; it sounds like a sensible *default*.